From: Konstantin Vasin <tempest921@gmail.com>
Date: Mon, 9 Mar 2020 17:41:22 +0300
Subject: [PATCH] netfilter: nf_flow_table_hw: check the status of
 dst_neigh

It's better to check the nud_state is VALID.
If there is not neigh previos, the lookup will
create a non NUD_VALID with 00:00:00:00:00:00 mac.

This fix is based on the upstream commit f31ad71c44
("netfilter: nf_flow_table_offload: check the status of dst_neigh")
from wenxu <wenxu@ucloud.cn>

Signed-off-by: Konstantin Vasin <tempest921@gmail.com>
---

index e831c8830e91..1238d675a316 100644
--- a/net/netfilter/nf_flow_table_hw.c
+++ b/net/netfilter/nf_flow_table_hw.c
@@ -44,8 +44,10 @@ static void flow_offload_check_ethernet(struct flow_offload *flow,
 	if (!n)
 		return;
 
-	memcpy(path->eth_dest, n->ha, ETH_ALEN);
-	path->flags |= FLOW_OFFLOAD_PATH_ETHERNET;
+	if (n->nud_state & NUD_VALID) {
+		memcpy(path->eth_dest, n->ha, ETH_ALEN);
+		path->flags |= FLOW_OFFLOAD_PATH_ETHERNET;
+	}
 	neigh_release(n);
 }
 
